torch/distributed/pipelining/stage.py

```diff
@@ -676,14 +675,19 @@ def backward_one_chunk(
                     "full", bwd_kwargs
                 )
             else:
-                # perform the partial backwards for the inputs with a custom backward function
-                # when the "stage_ouput" is a loss, then it is a tensor, otherwise it is a tuple of tensors
-                if isinstance(bwd_kwargs["stage_output"], torch.Tensor):
-                    bwd_kwargs["stage_output"] = (bwd_kwargs["stage_output"],)
-
-                grads_input, param_groups = self.backward_maybe_with_nosync(
-                    "input", bwd_kwargs
-                )
+                grads_input = []
+                param_groups = []
+                # Skip the backward for the first stage since we will perform the weight update with
```

Comment:
Does this mean that stage 0 will never run separate W/I computations even in multi stage pipelines? 

I think this is a significant problem since in ZB it is more common to use separate W/I for earlier stages than late stages. Last stage may have almost entirely merged full-backward but first stage may need mostly separated ones to fill bubbles.

Reply:
Stage 0 still computes W/I but now the I is like a no-op since the real work is in done in W. Typically the input grad would not be computed for stage 0 anyways since the inputs do not require gradients and this skips the `.grad()` call entirely.

This is only for the case of W/I split, for the full B the backward execution will remain the same.